cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,950 @@
+  //!@param inccarX stride within X's carry vector (use every inccarX'th element)
+  //!@param incpriY stride within Y's primary vector (use every incpriY'th element)
+  //!@param inccarY stride within Y's carry vector (use every inccarY'th element)
+  _CCCL_DEVICE void binned_dmdmadd(
+    const ReproducibleFloatingAccumulator& x, const int incpriX, const int inccarX, const int incpriY, const int inccarY)
+  {
+    if (is_zero_v(x.primary(0)))
+    {
+      return;
+    }
+
+    if (is_zero_v(primary(0)))
+    {
+      for (int i = 0; i < FOLD; i++)
+      {
+        primary(i * incpriY) = x.primary(i * incpriX);
+        carry(i * inccarY)   = x.carry(i * inccarX);
+      }
+      return;
+    }
+
+    if (is_nan_inf_v(x.primary(0)) || is_nan_inf_v(primary(0)))
+    {
+      primary(0) += x.primary(0);
+      return;
+    }
+
+    const auto X_index = x.binned_index();
+    const auto Y_index = this->binned_index();
+    const auto shift   = Y_index - X_index;
+    if (shift > 0)
+    {
+      // shift Y upwards and add X to Y
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = FOLD - 1; i >= 1; i--)
+      {
+        if (i < shift)
+        {
+          break;
+        }
+        primary(i * incpriY) =
+          x.primary(i * incpriX) + (primary((i - shift) * incpriY) - binned_bins(i - shift + Y_index));
+        carry(i * inccarY) = x.carry(i * inccarX) + carry((i - shift) * inccarY);
+      }
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = 0; i < FOLD; i++)
+      {
+        if (i == shift)
+        {
+          break;
+        }
+        primary(i * incpriY) = x.primary(i * incpriX);
+        carry(i * inccarY)   = x.carry(i * inccarX);
+      }
+    }
+    else if (shift < 0)
+    {
+      // shift X upwards and add X to Y
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = 0; i < FOLD; i++)
+      {
+        if (i < -shift)
+        {
+          continue;
+        }
+        primary(i * incpriY) += x.primary((i + shift) * incpriX) - binned_bins(X_index + i + shift);
+        carry(i * inccarY) += x.carry((i + shift) * inccarX);
+      }
+    }
+    else if (shift == 0)
+    {
+      // add X to Y
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = 0; i < FOLD; i++)
+      {
+        primary(i * incpriY) += x.primary(i * incpriX) - binned_bins(i + X_index);
+        carry(i * inccarY) += x.carry(i * inccarX);
+      }
+    }
+
+    binned_dmrenorm(incpriY, inccarY);
+  }
+
+  /// Add two manually specified binned fp (Y += X)
+  /// Performs the operation Y += X
+  _CCCL_DEVICE void binned_dbdbadd(const ReproducibleFloatingAccumulator& other)
+  {
+    binned_dmdmadd(other, 1, 1, 1, 1);
+  }
+
+public:
+  ReproducibleFloatingAccumulator()                                       = default;
+  ReproducibleFloatingAccumulator(const ReproducibleFloatingAccumulator&) = default;
+  /// Sets this binned fp equal to another binned fp
+  ReproducibleFloatingAccumulator& operator=(const ReproducibleFloatingAccumulator&) = default;
```

Comment:
Is the purpose to suppress move constructor / assignment?

Otherwise this can go